### PR TITLE
Relate non-augmenting subtypes without resorting to structural comparison

### LINKED ISSUE #43422: Perf regression: preserving union aliases
I'm seeing ~1.5s of check time in 0c58edeb294609202c9065861cec02b2d08cd922 and ~5s of check time in 6aeb8c12ccc72ecfbd1b3c227c63ec3412c06447 (aka #42149).

Note that the repro is really brittle - virtually any change causes the check time to drop dramatically (including: inlining the .d.ts contents in the .ts file, using a type alias for the array type, even dropping the `width` property).

The real `CSSObject` type is from https://github.com/emotion-js/emotion.

index.ts
```ts
import type { CSSObject } from "./emotion";

declare const width: string | number;
declare const y: CSSObject;

const x: CSSObject = {
    width,
    ...y,
};
```

emotion.d.ts
```ts
import * as CSS from 'csstype'

type CSSProperties = CSS.PropertiesFallback<number | string>
type CSSPropertiesWithMultiValues = {
  [K in keyof CSSProperties]:
    | CSSProperties[K]
    | Array<Exclude<CSSProperties[K], undefined>>
}

interface ArrayCSSInterpolation extends Array<CSSInterpolation> {}
// type ArrayCSSInterpolation = Array<CSSInterpolation>;

type InterpolationPrimitive =
  | null
  | undefined
  | boolean
  | number
  | string
  | CSSObject

type CSSInterpolation = InterpolationPrimitive | ArrayCSSInterpolation

interface CSSOthersObject {
  [propertiesName: string]: CSSInterpolation
}

export interface CSSObject extends CSSPropertiesWithMultiValues, CSSOthersObject {}
```

Note that you'll need to `npm install csstype`.

Command: `tsc index.ts -skipLibCheck -noEmit -extendedDiagnostics`

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #43422 (this improves the perf of the example given to better-than-4.1-or-3.6 performance on my machine)

Much like the inspiration for #43435, this PR is about introducing a fast path for comparing trivial subtypes of `Array<T>` with `Array<T>` itself, without resorting to a full, slow, structural comparison (which is _especially_ useful in cases where the comparison succeeds, as it avoids an exhaustive traversal of the `Array` type structure). This is useful, since non-augmenting array subtypes turn out to be pretty commonly used in the wild for describing recursive type structures in versions of TS that predate recursive type alias support.

Then, that's been generalized to normalize any empty subtype to it's single base type, if applicable, sharing these savings to empty subtypes of `Map` and other types, should they be present.

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot pack this
@typescript-bot perf test this

**Comment by weswigham** [comment]
@typescript-bot pack this
@typescript-bot perf test this

@amcasey once the tarball is done, do you have a local non-reduced repro to verify this on?

**Comment by amcasey** [comment]
@weswigham The actual repro is https://github.com/vanilla/vanilla.  You might get to it before I do, since I'm still on DT and not yet done with meetings for today.

**Comment by andrewbranch** [comment]
Why do the interfaces have to be non-augmenting? Checks on the interface declaration itself will ensure it’s assignable to its base type, so why can’t we just relate the type argument in the subtype interface’s `extends` clause to the target (`Array`) type parameter? E.g.

```ts
interface WhateverArray extends Array<string> {
  extraMember: unknown;
}

let x: (string | number)[] = undefined! as WhateverArray;
```

Wouldn’t it be sufficient here to check that `string` is assignable to `string | number`?

**Comment by weswigham** [comment]
Teeeechnically non-conforming "subtypes" of array can exist in programs with errors, eg:
```ts
interface MyBadArray extends Array<string> {
    pop(): void;
}

declare var x: MyBadArray;
declare var y: Array<string>;

x = y;
y = x;
```
so _relying_ on the subtype check might not be the best idea - especially if someone's `@ts-ignore`'d an error like this.

What I _could_ do is expand this to non-array subtypes, eg
```ts
interface MyMap extends Map<string, string> {}

declare var x: MyMap;
declare var y: Map<string, string>;

x = y;
y = x;
```
and instead just check if the single base type of the type with no augmentations is or is an instantiation of the base type or target of the base type.

**Comment by amcasey** [comment]
Timing for Vanilla is basically the same as https://github.com/microsoft/TypeScript/pull/43623#pullrequestreview-633972861 except that this gets things down to about 19s, rather than 22s.

I think both changes are good, even if they aren't both necessary to speed up Vanilla.

File: tests/baselines/reference/arrayLiterals3.errors.txt [hunk 1]
```diff
@@ -8,7 +8,7 @@ tests/cases/conformance/expressions/arrayLiterals/arrayLiterals3.ts(17,5): error
 tests/cases/conformance/expressions/arrayLiterals/arrayLiterals3.ts(33,5): error TS2322: Type 'number[]' is not assignable to type '[number, number, number]'.
   Target requires 3 element(s) but source may have fewer.
 tests/cases/conformance/expressions/arrayLiterals/arrayLiterals3.ts(34,5): error TS2322: Type '(string | number)[]' is not assignable to type 'myArray'.
-  The types returned by 'pop()' are incompatible between these types.
+  Type '(string | number)[]' is not assignable to type 'Number[]'.
```
Comment (amcasey) [inline tests/baselines/reference/arrayLiterals3.errors.txt:11]: Seems better

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -18219,6 +18219,22 @@ namespace ts {
                     return result;
                 }
 
+                if (isNonAugmentingArraySubtype(source) && isArrayType(target)) {
```
Comment (amcasey) [inline src/compiler/checker.ts:18222]: What if they're both non-augmenting array subtypes?

**Review by amcasey — APPROVED** [review]
Looks sensible and seems to work in practice, but I don't know enough to vouch for the correctness.

File: src/compiler/checker.ts (hunk 2 above)
Comment (weswigham) [reply src/compiler/checker.ts:18222]: We could relate them; it's just not as common in practice. This comes back to my "we could generalize this more" remark in my prior comment - this doesn't really need to be limited to just `Array`s either.

**Comment by weswigham** [comment]
@amcasey I assume this PR prevents the degenerate `Maybe` runaway whose perf is fixed in the other PR from ever occuring in the first place.

**Comment by amcasey** [comment]
@weswigham Yes, that would be my guess as well, but I figured there might be other pseudo-aliases or other maybe-stack wins.

**Comment by ahejlsberg** [comment]
I think it might be worthwhile generalizing this to all non-augmenting interface types with a single base type and also rolling it into `getNormalizedType`. It's not an uncommon pattern.

**Comment by DanielRosenwasser** [comment]
@ahejlsberg @weswigham this was one of the sorts of things that @amcasey and I had brought up when we mentioned that type aliases were actually faster than extending interfaces.

You can imagine a general fast path that when an interface

* has no call/construct/index signatures
* has base types with no call/construct/index signatures
* has properties that are mutually exclusive with those of their base types
* has base types whose property names are mutually exclusive with each other's

then you simply copy the members over and set their assignability cache key.

But keeping it simple with just extending one type and introducing no members is a good start too.

**Comment by andrewbranch** [comment]
> Teeeechnically non-conforming "subtypes" of array can exist in programs with errors.... so relying on the subtype check might not be the best idea... since check order is arbitrary

It still seems a little silly to me that we do the work of checking whether an interface properly extends its base types, and then do these expensive structural checks just in case it didn’t. Like, we have the ability to ensure that an interface declaration is checked against its base types before being used in an assignability check, right? (I don’t know how that wouldn’t be the case, since we’d need to resolve members from its base types in order to check assignability anyway.) Then we could flag interface types that _don’t_ properly extend their base types to ensure we fall back to expensive structural checks when we need to.

It seems like that would address the comment in #43623:

> But #43624 only fixes the issue for non-augmenting interface types. If the `ArrayCSSInterpolation` type in the repro were to add any members, the problem would be right back. So maybe with need both PRs.

But I’m sure I’m missing some important point here.

**Comment by DanielRosenwasser** [comment]
> we have the ability to ensure that an interface declaration is checked against its base types before being used in an assignability check, right

We do, I'm pretty sure

```
for (const baseType of getBaseTypes(type)) {
    checkTypeAssignableTo(typeWithThis, getTypeWithThisArgument(baseType, type.thisType), node.name, Diagnostics.Interface_0_incorrectly_extends_interface_1);
}
```

**Comment by weswigham** [comment]
@DanielRosenwasser that work itself is _expensive_ because it's _structural_ (without this PR). (Also it _is_ cached but maybe not as well as you think because of `this`-type shenanigans!)

**Comment by weswigham** [comment]
@ahejlsberg 

> I think it might be worthwhile generalizing this to all non-augmenting interface types with a single base type and also rolling it into getNormalizedType. It's not an uncommon pattern.

Done.

@typescript-bot run dt
@typescript-bot test this
@typescript-bot user test this
@typescript-bot perf test this
@typescript-bot pack this

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -19901,6 +19903,28 @@ namespace ts {
             return isArrayType(type) || !(type.flags & TypeFlags.Nullable) && isTypeAssignableTo(type, anyReadonlyArrayType);
         }
 
+        function getSingleBaseForNonAugmentingSubtypeOrType(type: Type) {
+            if (!(getObjectFlags(type) & ObjectFlags.Reference) || !(getObjectFlags((type as TypeReference).target) & ObjectFlags.ClassOrInterface)) {
+                return type;
+            }
+            if ((type as TypeReference).cachedEquivalentBaseTypeOrSelf) {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts:19910]: Looks like we take close to a 1% perf hit from adding this property to every type reference. Maybe use a pair of `ObjectFlags` to track whether (a) you've computed this information and (b) whether there is an equivalent base type, and only add the property then. You could co-opt the `IsNeverIntersectionComputed` and `IsNeverIntersection` bit positions for this use in type references as they're currently used only in intersections.

BTW, I also notice that someone added an `ObjectFlags.IsClassInstanceClone` with the value `1 << 30`. That's outside the small integer range and means the VM will resort to boxing and de-optimizing whenever that flags is set.

**Review by ahejlsberg — CHANGES_REQUESTED** [review]
I really like this change, but let's see if we can get the perf cost down.

File: src/compiler/checker.ts (hunk 3 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts:19910]: I just put up #43732 which reorganizes the `ObjectFlags` enum. Should be pretty easy now to add the flags I suggest above.

**Comment by weswigham** [comment]
@typescript-bot perf test this

File: src/compiler/checker.ts (hunk 3 above)
Comment (weswigham) [reply src/compiler/checker.ts]: Done; but it looks like my comment on https://github.com/microsoft/TypeScript/pull/43732#pullrequestreview-639191529 was warranted, though - I had to fix a crash reusing those flags triggered. Should we audit our other `ObjectFlags` usages of these type-dependent flags to ensure there aren't other crashes that simply aren't triggered by our test suite?

Comment (weswigham) [reply src/compiler/checker.ts]: I did a quick audit and found one more location we were missing a `TypeFlags` check - I've opened https://github.com/microsoft/TypeScript/pull/43745 with just those changes.

**Comment by weswigham** [comment]
@ahejlsberg this now uses flags to indicate the presence of a cache value, as requested.

**Comment by weswigham** [comment]
@ahejlsberg ping?

File: src/compiler/types.ts [hunk 4]
```diff
@@ -5220,6 +5220,12 @@ namespace ts {
         IsNeverIntersectionComputed = 1 << 26, // IsNeverLike flag has been computed
         /* @internal */
         IsNeverIntersection = 1 << 27, // Intersection reduces to never
+
+        // Flags that require TypeFlags.Reference
```
Comment (ahejlsberg) [inline src/compiler/types.ts:5224]: There is no `TypeFlags.Reference`. Move them to the section that requires `TypeFlags.Object` and use bits 25 and 26.

**Review by ahejlsberg — APPROVED** [review]
Approved with the minor change below.

**Comment by xaviergonz** [comment]
Just FYI: I think this PR seems to be related to this bug
https://github.com/microsoft/TypeScript/issues/44281

**Pull request merged**